README.md
# Kodi Hyperion Addon

##### Install Guide
- download and install Kodi repository: http://kodi.z-index.net/addons/repository.zindex.zip
- install Hyperion Addon from my repository

German support thread: http://powerpi.de/forum/viewtopic.php?t=190


changelog.txt
version 0.0.6
 - switch from hyperion remote-binary to json interface
 - add more log

version 0.0.5
 - new effects

version 0.0.4
 - switch additional executable

version 0.0.3
 - small bug fixes
 - code improvements

version 0.0.2
 - working service interface
 - enhanced json rpc interface

version 0.0.1
  - initial release
  - working json rpc interface for color/effect/clear commands


service.py
import xbmc
import xbmcaddon
import xbmcgui

from resources.lib import hyperion

__addon__ = xbmcaddon.Addon()
__title__ = __addon__.getAddonInfo('name')
__icon__ = __addon__.getAddonInfo('icon')
__language__ = __addon__.getLocalizedString
__hyperion__ = None


def log(txt):
    message = '%s: %s' % (__title__, txt.encode('ascii', 'ignore'))
    xbmc.log(msg=message, level=xbmc.LOGNOTICE)


class Main:
    def __init__(self):
        hyperion_host = __addon__.getSetting('hyperion_host')
        hyperion_port = __addon__.getSetting('hyperion_port')
        priority = __addon__.getSetting('priority')
        global __hyperion__

        if not hyperion_host or not hyperion_port:
            log('Service: ' + __language__(20000))
            xbmcgui.Dialog().notification(__title__, __language__(20000), __icon__, 5000)
            return

        __hyperion__ = hyperion.Remote(hyperion_host=hyperion_host, hyperion_port=hyperion_port, priority=priority)

        self.Player = MyPlayer()
        self.Monitor = MyMonitor()
        self._daemon()

    def _daemon(self):
        while not self.Monitor.abortRequested():
            if self.Monitor.waitForAbort(10):
                self.Monitor.onShutdown()
                break


class MyMonitor(xbmc.Monitor):
    def __init__(self):
        xbmc.Monitor.__init__(self)
        self.onStart()

    def onStart(self):
        if __addon__.getSetting('clear_on_start') == 'true':
            log('Monitor: onStart(clearAll)')
            __hyperion__.clearAll()

        if __addon__.getSetting('effect_on_kodi_startup'):
            log('Monitor: onStart(effect)')
            __hyperion__.effect(__addon__.getSetting('effect_on_kodi_startup'))

    def onShutdown(self):
        if __addon__.getSetting('off_on_shutdown') == 'true':
            log('Monitor: onShutdown')
            __hyperion__.color('black')

    def onScreensaverActivated(self):
        if __addon__.getSetting('off_on_screensaver_activated') == 'true':
            log('Monitor: onScreensaverActivated')
            __hyperion__.color('black')

    def onScreensaverDeactivated(self):
        if __addon__.getSetting('effect_on_screensaver_deactived'):
            log('Monitor: onScreensaverDeactivated')
            __hyperion__.effect(__addon__.getSetting('effect_on_screensaver_deactived'))


class MyPlayer(xbmc.Player):
    def __init__(self):
        xbmc.Player.__init__(self)

    def onPlayBackStarted(self):
        if __addon__.getSetting('clear_on_video_playback') and xbmc.Player().isPlayingVideo():
            log('Player: onPlayBackStarted(clearAll)')
            __hyperion__.clearAll()

        if __addon__.getSetting('effect_on_playback_audio') and xbmc.Player().isPlayingAudio():
            log('Player: onPlayBackStarted(effect)')
            __hyperion__.effect(__addon__.getSetting('effect_on_playback_audio'))

    def onPlayBackResumed(self):
        if __addon__.getSetting('clear_on_video_playback') == 'true' and xbmc.Player().isPlayingVideo():
            log('Player: onPlayBackResumed')
            __hyperion__.clearAll()

    def onPlayBackPaused(self):
        if __addon__.getSetting('effect_on_playback_paused') and xbmc.Player().isPlayingVideo():
            log('Player: onPlayBackPaused')
            __hyperion__.effect(__addon__.getSetting('effect_on_playback_paused'))

    def onPlayBackStopped(self):
        if __addon__.getSetting('effect_on_playback_stopped'):
            log('Player: onPlayBackStopped')
            __hyperion__.effect(__addon__.getSetting('effect_on_playback_stopped'))

    def onPlayBackEnded(self):
        self.onPlayBackStopped()


if __name__ == '__main__':
    if __addon__.getSetting('autostart') == 'true':
        log('Service: autostart execute')
        Main()


resources/lib/hyperion.py
import json
import socket

from resources.lib import webcolors


class Remote:
    def __init__(self, hyperion_host='127.0.0.1', hyperion_port='19444', priority='0'):
        self.hyperion_host = hyperion_host
        self.hyperion_port = hyperion_port
        self.priority = priority

        try:
            import StorageServer
        except:
            from resources.lib import storageserverdummy as StorageServer

        self.cache = StorageServer.StorageServer('hyperion', 8544)

    def color(self, color, priority=None):

        payload = {'command': 'color', 'color': webcolors.name_to_rgb(color), 'priority': priority}

        if color == 'black':
            self.setState('off')
        else:
            self.setState('on')

        return self.run(payload)

    def effect(self, effect, priority=None):
        payload = {'command': 'effect', 'effect': {'name': effect}, 'priority': priority}

        self.setState('on')

        return self.run(payload)

    def clear(self, priority=None):
        payload = {'command': 'clear', 'priority': priority}

        self.setState('on')

        return self.run(payload)

    def clearAll(self):
        payload = {'command': 'clearall'}
        self.setState('on')

        return self.run(payload)

    def serverinfo(self):
        payload = {'command': 'serverinfo'}

        return self.run(payload)

    def run(self, payload):

        if not payload:
            return False

        if 'priority' in payload and payload['priority'] is None:
            payload['priority'] = int(self.priority)

        data = json.dumps(payload) + '\n'

        try:
            ret = self.nc(data)
        except Exception, e:
            print "Hyperion.Remote: " + str(e)
            return False

        print 'Hyperion.Remote: ' + str(ret)

        return ret

    def setState(self, state):
        self.cache.set('state', state)

    def getState(self):
        return self.cache.get('state')

    def nc(self, data):
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

            s.connect((self.hyperion_host, int(self.hyperion_port)))

            s.sendall(data)

            data = s.recv(4096)

            s.close()

        except socket.error as exc:
            raise Exception(exc)

        return json.loads(data)


default.py
import sys
import subprocess
import xbmc
import xbmcaddon
import xbmcgui

from urlparse import parse_qsl
from resources.lib import hyperion

__addon__ = xbmcaddon.Addon()
__title__ = __addon__.getAddonInfo('name')
__icon__ = __addon__.getAddonInfo('icon')
__language__ = __addon__.getLocalizedString
__hyperion__ = None


def log(txt):
    message = '%s: %s' % (__title__, txt.encode('ascii', 'ignore'))
    xbmc.log(msg=message, level=xbmc.LOGNOTICE)


class Main:
    def __init__(self):
        hyperion_host = __addon__.getSetting('hyperion_host')
        hyperion_port = __addon__.getSetting('hyperion_port')
        priority = __addon__.getSetting('priority')

        global __hyperion__

        if not hyperion_host or not hyperion_port:
            print 'Hyperion.Script: ' + __language__(20000)
            xbmcgui.Dialog().notification(__title__, __language__(20000), __icon__, 5000)
            return

        __hyperion__ = hyperion.Remote(hyperion_host=hyperion_host, hyperion_port=hyperion_port, priority=priority)

        self.params = dict(parse_qsl('&'.join(sys.argv)))

        if not self.params:
            xbmc.executebuiltin('Addon.OpenSettings(script.hyperion)')

        command = self.params.get('command', None)

        log('Script: command: %s' % command)

        if command is None:
            log('Script: no command given?!')

        elif command == 'effect':
            self._effect()

        elif command == 'color':
            self._color()

        elif command == 'clear':
            self._clear()

        elif command == 'clearall':
            self._clearAll()

        elif command == 'switch':
            self._switch()

    def _effect(self):
        effect = self.params.get('effect', None)
        priority = self.params.get('priority', None)

        if not effect:
            return

        xbmcgui.Dialog().notification(__title__, effect, __icon__, 3000)
        __hyperion__.effect(effect, priority)

    def _color(self):
        color = self.params.get('color', None)
        priority = self.params.get('priority', None)

        if not color:
            return

        try:
            __hyperion__.color(color, priority)
        except ValueError as e:
            color = e.message

        xbmcgui.Dialog().notification(__title__, color, __icon__, 3000)

    def _clear(self):
        priority = self.params.get('priority', None)

        xbmcgui.Dialog().notification(__title__, __language__(20001), __icon__, 3000)
        __hyperion__.clear(priority)

    def _clearAll(self):
        xbmcgui.Dialog().notification(__title__, __language__(20001), __icon__, 3000)
        __hyperion__.clearAll()

    def _switch(self):

        if not __addon__.getSetting('switch_type'):
            return

        priority = self.params.get('priority', None)

        if __hyperion__.getState() == 'on':
            __hyperion__.color('black', priority)
        else:
            if __addon__.getSetting('switch_type') == '1':
                __hyperion__.effect(__addon__.getSetting('switch_effect'), priority)
            else:
                __hyperion__.clearAll()

        if __addon__.getSetting('switch_additional'):
            cmd = '%s %s' % (__addon__.getSetting('switch_additional'), __hyperion__.getState())
            log('Script: switch additional %s' % cmd)
            subprocess.Popen(cmd, shell=True)


if __name__ == '__main__':
    log('Script: execute')
    Main()
